icepyx/core/harmony.py

```diff
@@ -1,13 +1,62 @@
-from typing import Any
+import datetime as dt
+from typing import Any, NotRequired, TypedDict, Union
 
-import requests
+import harmony
 
-from icepyx.core.urls import CAPABILITIES_BASE_URL
+from icepyx.core.auth import EarthdataAuthMixin
 
 
-def get_capabilities(concept_id: str) -> dict[str, Any]:
-    response = requests.get(
-        CAPABILITIES_BASE_URL,
-        params={"collectionId": concept_id},
-    )
-    return response.json()
+class HarmonyTemporal(TypedDict):
+    start: NotRequired[dt.datetime]
+    stop: NotRequired[dt.datetime]
+
+
+class HarmonyApi(EarthdataAuthMixin):
+    def __init__(self):
+        # initialize authentication properties
+        EarthdataAuthMixin.__init__(self)
+        self.harmony_client = harmony.Client(
+            auth=(
+                self.auth.username,
+                self.auth.password,
+            ),
+        )
+
+    def get_capabilities(self, concept_id: str) -> dict[str, Any]:
+        capabilities_request = harmony.CapabilitiesRequest(concept_id=concept_id)
+        response = self.harmony_client.submit(capabilities_request)
+
+        return response
+
+    def place_order(
+        self,
+        concept_id: str,
+        # These are optional subset parameters
+        bounding_box: Union[harmony.BBox, None] = None,
+        temporal: Union[HarmonyTemporal, None] = None,
+    ) -> str:
+        """Places a Harmony order with the given parameters.
+
+        Return a string representing a job ID.
+
+        TODO/Question: it looks like this code will always use the provided
+        parameters to do subsetting. Are there cases where we just want the data
+        downloaded as whole granules? If so, we may need to use another API to
+        do so?
```

Comment:
Interesting. The current icepyx default is to subset automatically for spatial and temporal (and then users could add any other available subsetting/reformatting options - it's probably best explained in the example notebooks for access and subsetting). In ipx.query, the order/download functions accept `subset=False` (would want to verify I have the kwarg right), so that would probably be the best way to trigger a non-subsetted download. I'm looking at transitioning that functionality over to earthaccess (with the ultimate goal of retiring ipx.Query entirely in favor of earthaccess), so we can work this case in there.

Reply:
Thanks for the feedback! I'm planning on using earthaccess to fill the gaps that harmony can't handle (non-subset orders).

LOTS more changes incoming :) Note that you don't need to keep up with this PR. I'm really just "going at it" and learning/breaking a lot as I go. My primary goal right now is assessing how much work it is going to be for us to implement harmony support, and we may need to revisit (or completely re-write) large chunks later.